### PR TITLE
[lsp] Add vscode plugin with goal display support

### PR DESCRIPTION
**Pull request by houdaM97**
Hi, 
this is the last version of my work on vscode extension for lambda pi (LSP client/server), it still to upload merge and upload to market place

### CODE REVIEW

File: src/proof.ml [hunk 1]
```diff
@@ -73,30 +73,33 @@ let focus_goal : proof_state -> Env.t * term = fun ps ->
 let pp_goals : _ pp = fun oc gl ->
```
Comment (rlepigre) [inline src/proof.ml]: This function is also used for printing in the console, so please revert the changes to this file. If you can use arbitrary HTML, why not use the format ``"<pre>%a</pre>"`` when printing with the (unchanged) function?

**Comment by fblanqui** [comment]
Ok, do you mind to adapt the code @rlepigre ?

**Comment by rlepigre** [comment]
> Ok, do you mind to adapt the code @rlepigre ?

Fixed. That was trivial.

File: src/proof.ml (hunk 1 above)
Comment (fblanqui) [reply src/proof.ml]: Solved in 9b99f85602779c9d988250295732f7f4f61f1a80.

File: src/proof.ml [hunk 2]
```diff
@@ -94,7 +94,8 @@ let pp_goals : _ pp = fun oc gl ->
           let (_, a) = Goal.get_type g in
           Format.fprintf oc "%i. %a\n" (i+1) pp a
         in
-        List.iteri print_goal gs
+        List.iteri print_goal gs;
+        Format.fprintf oc "\n"
```
Comment (rlepigre) [inline src/proof.ml]: Please, revert to previous state (newline is not useful here).

**Comment by fblanqui** [comment]
@ejgallego , @rlepigre , do you agree to merge this PR?

**Comment by ejgallego** [comment]
No, it should not be merged as-is; it needs a few cleanups and more importantly the removal of a lot of commits that don't belong here. I will take care soon, I just returned from holidays.

**Comment by fblanqui** [comment]
Hi @ejgallego . Do you think you can do this soon? @Rehan-MALAK just started his postdoc and needs to work on vscode.

**Comment by ejgallego** [comment]
I'll be back next week, will that work?

**Comment by Rehan-MALAK** [comment]
Cool !
Meanwhile, could somebody give me a combination of a lambdapi commit + vscode version + nodejs/npm version where we are sure that the basics of the plugins are working (ie. : show goal).

**Comment by ejgallego** [comment]
The branch here should work.

File: editors/vscode/src/client.ts [hunk 3]
```diff
@@ -32,7 +31,6 @@ let inset : WebviewEditorInset;
 export function activate(context: ExtensionContext) {
 
 	//create the server
-
 	let serverOptions = {
 		command: '/Users/houdamouzoun/.opam/4.05.0/bin/lp-lsp',
```
Comment (Rehan-MALAK) [inline editors/vscode/src/client.ts]: hardcoded user-specific path

**Comment by fblanqui** [comment]
@Rehan-MALAK don't hesitate to modify this PR or create a new one to fix it.

File: editors/vscode/src/client.ts (hunk 3 above)
Comment (fblanqui) [reply editors/vscode/src/client.ts]: Fixed in https://github.com/Deducteam/lambdapi/pull/240/commits/65b60d11c15cfdb5b737e94f5cf6b3bd14060d71

**Comment by fblanqui** [comment]
How do I use npx tsc instead of tsc?

**Comment by Rehan-MALAK** [comment]
> How do I use npx tsc instead of tsc?

you could substitute ``tsc`` by ``npx tsc`` in package.json but I dont know the proper way to do it

if you choose the local installation of tsc, I would just put tsc directory in the PATH

**Comment by ejgallego** [comment]
Almost done, my rebase does conflict with some of the changes here, I suggest I push the updated version here, then you folks rebase #268 ; WDYT?

**Comment by Rehan-MALAK** [comment]
I havn't pushed anything here and the 6 commits in #268 are just a desperate try to compile the plugin from last week so it will be easy to rebase or to drop them => #268 will be closed.
But if you have conflicts, you are probably the most knowledgeable on the necessity to keep or not @fblanqui and @rlepigre commits 83b6759 9b99f85 71ec560 9dfd0ed 2be0db7 .

**Comment by fblanqui** [comment]
https://github.com/Deducteam/lambdapi/commit/83b67592c660d5be510011081daa4a628e621c02 is a merge. I don't know what it is about.
https://github.com/Deducteam/lambdapi/commit/71ec5601566092385937b54dc5d4c3e5340a750f splits long lines into smaller ones only.
https://github.com/Deducteam/lambdapi/commit/9b99f85602779c9d988250295732f7f4f61f1a80 is important and should be kept.

**Comment by ejgallego** [comment]
Ok I am close to finish this, I hit some problems as I was using vscode 1.40, I had to go back to vscode 1.37 ummm, that was a pain :)

@fblanqui https://github.com/Deducteam/lambdapi/commit/9b99f85602779c9d988250295732f7f4f61f1a80 is already into my tree as I think the LSP server wants to use a custom print function anyways, for example, to insert commands.

**Comment by ejgallego** [comment]
Now I am getting some other messages related to the handling of modules; I guess due to changes in `master`; weird...

**Comment by Rehan-MALAK** [comment]
I had recently some problems with the upgrade on Debian from code_1.39.2-1571154070 to code_1.40.0-1573064499.
I dont know exactly why so I removed :
~/node_modules
~/.config/Code
~/.vscode
and nodejs package and resinstalled everything.
I hope we will be able to make it work with latest vscode (currently 1.40) and ProposedAPI

**Comment by ejgallego** [comment]
@Rehan-MALAK I will investigate more; I've pushed a functional version [using 1.37] , this is still not ready to merge tho, it is missing a patch from Houda and patch from mine, but you can try testing.

I have incorporated some of your changes [regarding gitignore and removing bad files], so I propose:
- you test this branch with 1.37 [note that goal display only works for incomplete proofs, but next patch by Houda will solve it]
- I push the two remaining patches to this branch
- we merge, and investigate how to make this work in 1.40 later

**Comment by Rehan-MALAK** [comment]
so luckily, the previous versions are still on the official website https://code.visualstudio.com/updates/v1_37

but how do you install the plugin, I added a Makefile in #268

currently, I can't even parse the json file

**Comment by Rehan-MALAK** [comment]
ok, it seems that comments are allowed or not in json files, depending on the platform, if I remove the comments, I can parse package.json

and the plugin is working with code 1.37 and  ``npm install`` and ``npm run compile`` with the goals diplayed for incomplete proofs ... Nice !

**Comment by ejgallego** [comment]
> ok, it seems that comments are allowed or not in json files, depending on the platform.. when I remove the comments, I can parse package.json
> 
> and the plugin is working with code 1.37 and `npm install` and `npm run compile` with the goals diplayed for incomplete proofs ... Nice !

Yay!! Sorry it took me so long @Rehan-MALAK ; I think this is ready for merge, I've pushed a new version and hopefully it should be OK.

You may want to port some changes from your branch; I will see what's wrong with 1.38 etc....

**Comment by ejgallego** [comment]
@fblanqui @rlepigre please review and merge when you are satisfied.

File: editors/vscode/.vscode/launch.json [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      }
+    },
+    {
+      "type": "node",
+      "request": "attach",
+      "name": "Attach to Server",
+      "port": 6009,
+      "restart": true,
+      //"outFiles": ["${workspaceRoot}/server/out/**/*.js"]
+      "outFiles": ["${workspaceRoot}/../../lsp-lsp/lp_lsp.ml"]
```
Comment (fblanqui) [inline editors/vscode/.vscode/launch.json]: Isn't it lp-lsp/ ?

File: editors/vscode/.vscode/launch.json [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	"script": "watch"
+      }
+    },
+    {
+      "type": "node",
+      "request": "attach",
+      "name": "Attach to Server",
+      "port": 6009,
+      "restart": true,
+      //"outFiles": ["${workspaceRoot}/server/out/**/*.js"]
```
Comment (fblanqui) [inline editors/vscode/.vscode/launch.json]: Why keeping this commented code?

File: editors/vscode/src/client.ts [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+// Returns the HTML code of the panel and the inset ccontent
+function buildGoalsContent(goals : Goal[]) {
+    
+    let header, footer : String;
+
+    // get the HTML code of goals environment
+    let codeEnvGoals : String = getGoalsEnvContent(goals);
+
+    // #FA8072
```
Comment (fblanqui) [inline editors/vscode/src/client.ts]: keep this comment?

File: lp-lsp/lp_lsp.ml [hunk 7]
```diff
@@ -29,8 +29,8 @@ module List = struct
 end
 
 module Option = struct
```
Comment (fblanqui) [inline lp-lsp/lp_lsp.ml]: Why not reusing Extra.Option instead?

File: editors/vscode/src/client.ts [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+    header =  `<!DOCTYPE html>
+￼       <html lang="en">
+￼       <head>
+￼               <meta charset="UTF-8">
+￼               <link rel="stylesheet" type="text/css" href="style/style.css" >
```
Comment (fblanqui) [inline editors/vscode/src/client.ts]: There seems to be no file style/style.css currently...

File: lp-lsp/lp_lsp.ml (hunk 7 above)
Comment (fblanqui) [reply lp-lsp/lp_lsp.ml]: This module is not used anymore. Remove it?

File: lp-lsp/lp_lsp.ml [hunk 9]
```diff
@@ -175,16 +187,103 @@ let get_goals ~doc ~line ~pos =
...
+  let goals = get_goals ~doc ~line ~pos in
+  let result = LSP.json_of_goals goals in
+  let msg = LSP.mk_reply ~id ~result in
+  LIO.send_json ofmt msg
+
+let myfail (msg : string) =
+  LIO.log_error "fail" msg;
+  failwith msg
+
+let get_token tokens pos =
```
Comment (fblanqui) [inline lp-lsp/lp_lsp.ml]: Some comment explaining what this function does would be welcome.

File: editors/vscode/.vscode/launch.json (hunk 4 above)
Comment (ejgallego) [reply editors/vscode/.vscode/launch.json]: Indeed this configuration is not useful; this is only supported for servers in Javascript / Typescript I think [can OCaml be debugged from vscode?].

I will remove it.

**Comment by ejgallego** [comment]
Updated `launch.json`.

File: editors/vscode/src/client.ts (hunk 8 above)
Comment (ejgallego) [reply editors/vscode/src/client.ts]: Umm maybe it was lost? Let me see.

File: lp-lsp/lp_lsp.ml (hunk 7 above)
Comment (ejgallego) [reply lp-lsp/lp_lsp.ml]: We can remove it if you want, but it tends to re-appear pretty often.

File: editors/vscode/src/client.ts (hunk 6 above)
Comment (ejgallego) [reply editors/vscode/src/client.ts]: Updated.

File: editors/vscode/src/client.ts (hunk 8 above)
Comment (ejgallego) [reply editors/vscode/src/client.ts]: In fact it never worked I think, so this is a todo, we didn't have time to figure out where it should be placed. Added a comment.

File: lp-lsp/lp_lsp.ml (hunk 9 above)
Comment (ejgallego) [reply lp-lsp/lp_lsp.ml]: I've done what I could, the function should likely be refactored in full.

**Comment by ejgallego** [comment]
I've confirmed the plugin works in 1.37-1.39 , however in vscode 1.40 the debugger cannot start the extension [maybe it works standalone]

No obvious reason, this deserves more investigation.

**Comment by ejgallego** [comment]
Identified the issue with 1.40 https://github.com/microsoft/vscode/issues/84685

**Comment by fblanqui** [comment]
@ejgallego let me know what you want me to merge.

**Comment by fblanqui** [comment]
Once this PR is merged, shouldn't we close #223, #268 and #269 as well?

**Comment by ejgallego** [comment]
> @ejgallego let me know what you want me to merge.

This is ready, please merge.

> Once this PR is merged, shouldn't we close #223, #268 and #269 as well?

- #223: let me take care of it next week, I think there are some tweaks I'd still like to do
- #268 includes some useful bits beyond this PR, so I guess @Rehan-MALAK could rebase on top of this one and still merge these bits.
- for #269 indeed in my opinion we can close; reasons are stated there but basically the LSP server architecture should become enough different from the command line client as the merge not worth it.

**Comment by fblanqui** [comment]
Thanks @ejgallego !

**Comment by ejgallego** [comment]
Thanks to you folks! Let me know of any feedback; I think the code is reasonably clear so it should be easy to tweak if we want. I am working on some incremental checking stuff will ping when is ready.
